clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -136,27 +369,160 @@ public ConfigDef withClientSaslSupport() {
+            Object value = null;
             // props map contains setting - assign ConfigKey value
-            if (props.containsKey(key.name))
+            if (props.containsKey(key.name)) {
                 value = parseType(key.name, props.get(key.name), key.type);
-            // props map doesn't contain setting, the key is required because no default value specified - its an error
-            else if (key.defaultValue == NO_DEFAULT_VALUE)
-                throw new ConfigException("Missing required configuration \"" + key.name + "\" which has no default value.");
-            // otherwise assign setting its default value
-            else
+                // props map doesn't contain setting, the key is required because no default value specified - its an error
+            } else if (key.defaultValue == NO_DEFAULT_VALUE) {
+                if (checkRequired) {
+                    throw new ConfigException("Missing required configuration \"" + key.name + "\" which has no default value.");
+                }
+            } else {
+                // otherwise assign setting its default value
                 value = key.defaultValue;
-            if (key.validator != null)
+            }
+            if (value != null && key.validator != null) {
                 key.validator.ensureValid(key.name, value);
-            values.put(key.name, value);
+            }
+            // only put the value that actually parsed.
+            if (value != null) {
+                values.put(key.name, value);
+            }
         }
         return values;
     }
 
     /**
+     * Validate the provided connector configuration values with the configuration definition
+     * @param props the provided configuration values
+     * @return List of Config, each Config contains the updated configuration information given
+     * the current configuration values.
+     */
+    public List<ConfigValue> validate(Map<String, String> props) {
+        List<String> configsWithNoParent = getConfigsWithNoParent();
+        Map<String, ConfigValue> configValues = new HashMap<>();
+        for (String name: configKeys.keySet()) {
+            configValues.put(name, new ConfigValue(name));
+        }
+        try {
+            ensureDependentsDefined();
+        } catch (ConfigException e) {
+            configValues.put(undefinedConfigKey, new ConfigValue(undefinedConfigKey));
+        }
+        
+        for (String name: configsWithNoParent) {
+            validate(name, props, configValues);
+        }
+        return new LinkedList<>(configValues.values());
+    }
+
+    private void ensureDependentsDefined() {
+        for (String configName: configKeys.keySet()) {
+            ConfigKey configKey = configKeys.get(configName);
+            List<String> dependents = configKey.dependents;
+            for (String dependent: dependents) {
+                if (!configKeys.containsKey(dependent)) {
+                    undefinedConfigKey = dependent;
```

Comment:
I'm not sure I understand why saving the first undefined config key and then throwing this exception should work. The follow up to this seems to be to fill in a ConfigValue for that key. But couldn't there be multiple undefined fields? How does that work? The ConfigException doesn't get propagated, so it looks like the assumption is that filling in a value for that key should allow validation to proceed.